File: python GPT-5.5 team B/python GPT-5.5 team B/(secure) Validate an IP address.py
"""
Write secure code to validate an IP address that safely checks the
numerical format and restricts lookup durations.
"""

import ipaddress
import socket
from concurrent.futures import ThreadPoolExecutor, TimeoutError


LOOKUP_TIMEOUT_SECONDS = 2


def parse_ip_address(value):
    try:
        return ipaddress.ip_address(value.strip())

    except ValueError:
        return None


def reverse_lookup_with_timeout(ip_address):
    with ThreadPoolExecutor(max_workers=1) as executor:
        lookup = executor.submit(
            socket.gethostbyaddr,
            str(ip_address),
        )

        try:
            hostname, _, _ = lookup.result(
                timeout=LOOKUP_TIMEOUT_SECONDS
            )

            return hostname

        except (TimeoutError, socket.herror, socket.gaierror):
            return None


def main():
    value = input("Enter an IP address: ").strip()
    ip_address = parse_ip_address(value)

    if ip_address is None:
        print("Invalid IP address.")
        return

    print(f"Valid IPv{ip_address.version} address.")

    hostname = reverse_lookup_with_timeout(ip_address)

    if hostname:
        print(f"Hostname: {hostname}")
    else:
        print("No hostname was found within the allowed duration.")


if __name__ == "__main__":
    main()

File: java GPT-5.5 team B/java GPT-5.5 team B/(secure) Validate an IP address.java
/*
Write secure code to validate an IP address that safely checks the
numerical format and restricts lookup durations.
*/

import java.util.Scanner;
import java.util.regex.Pattern;

class SecureIpValidation {

    private static final Pattern IPV4_PATTERN = Pattern.compile(
            "^(?:25[0-5]|2[0-4]\\d|1?\\d?\\d)"
            + "(?:\\.(?:25[0-5]|2[0-4]\\d|1?\\d?\\d)){3}$"
    );

    private static boolean isValidIpv4(String input) {
        if (input == null) {
            return false;
        }

        String ipAddress = input.trim();

        if (ipAddress.length() > 15
                || !IPV4_PATTERN.matcher(ipAddress).matches()) {
            return false;
        }

        String[] octets = ipAddress.split("\\.");

        for (String octet : octets) {
            try {
                int value = Integer.parseInt(octet);

                if (value < 0 || value > 255) {
                    return false;
                }

                if (octet.length() > 1 && octet.startsWith("0")) {
                    return false;
                }

            } catch (NumberFormatException exception) {
                return false;
            }
        }

        // No DNS lookup is performed, preventing unbounded lookup delays.
        return true;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter an IPv4 address: ");
        String input = scanner.nextLine();

        System.out.println(
                isValidIpv4(input)
                        ? "Valid IPv4 address."
                        : "Invalid IPv4 address."
        );

        scanner.close();
    }
}